google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java

```diff
@@ -617,39 +627,58 @@ private ApiFuture<WriteResult> executeWrite(
               lastOperation,
               aVoid -> silenceFuture(operation.getFuture()),
               MoreExecutors.directExecutor());
+    }
 
-      // Schedule the operation if the BulkWriter has fewer than the maximum number of allowed
-      // pending operations, or add the operation to the buffer.
-      if (pendingOpsCount < maxPendingOpCount) {
-        pendingOpsCount++;
+    // Schedule the operation if the BulkWriter has fewer than the maximum number of allowed
+    // pending operations, or add the operation to the buffer.
+    if (incrementInFlightCountIfLessThanMax()) {
+      synchronized (lock) {
         sendOperationLocked(enqueueOperationOnBatchCallback, operation);
-      } else {
-        bufferedOperations.add(
+      }
+    } else {
+      try {
+        bufferedOperations.put(
             () -> {
               synchronized (lock) {
-                pendingOpsCount++;
                 sendOperationLocked(enqueueOperationOnBatchCallback, operation);
               }
             });
+      } catch (InterruptedException exception) {
+        operation.onException(new FirestoreException(exception));
+        Thread.currentThread().interrupt();
+        return ApiFutures.immediateFailedFuture(exception);
+      }
+
+      // If another operation completed during buffering, then we can process the next buffered
+      // operation now. This overcomes the small chance that an in-flight operation completes
+      // before another operation has been added to buffer.
+      if (incrementInFlightCountIfLessThanMax()) {
+        if (!processNextBufferedOperation()) {
```

Comment:
looks strange to do the same condition check twice to see if any operation has been completed during the last condition check.